File: packages/js/client/src/createWeb3ApiClient.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable prefer-const */

import { Web3ApiClient, Web3ApiClientConfig } from "./Web3ApiClient";
import { PluginConfigs, modules, uris } from "./pluginConfigs";

import { PluginRegistration } from "@web3api/core-js";
import { Tracer } from "@web3api/tracing-js";

export { PluginConfigs };

export const createWeb3ApiClient = Tracer.traceFunc(
  "createWeb3ApiClient",
  async (
    pluginConfigs: PluginConfigs,
    config?: Partial<Web3ApiClientConfig>
  ): Promise<Web3ApiClient> => {
    const plugins: PluginRegistration[] = [];

    for (const plugin of Object.keys(pluginConfigs)) {
      let pluginModule: any;

      if (!modules[plugin]) {
        throw Error(
          `Requested plugin "${plugin}" is not a supported createWeb3ApiClient plugin.`
        );
      }

      try {
        pluginModule = await import(modules[plugin]);
      } catch (err) {
        throw Error(
          `Failed to import plugin module. Please install the package "${modules[plugin]}".\n` +
            `Error: ${err.message}`
        );
      }

      const pluginFactory = pluginModule["plugin"];

      if (!pluginFactory) {
        throw Error(
          `Plugin module "${modules[plugin]}" is missing the "plugin: PluginFactory" export.`
        );
      }

      if (typeof pluginFactory !== "function") {
        throw Error(
          `The "plugin: PluginFactory" export must be a function. Found in module "${modules[plugin]}".`
        );
      }

      const pluginPackage = pluginFactory(
        (pluginConfigs as Record<string, unknown>)[plugin]
      );

      if (
        !pluginPackage ||
        typeof pluginPackage !== "object" ||
        !pluginPackage.factory ||
        !pluginPackage.manifest
      ) {
        throw Error(
          `Plugin package is malformed. Expected object with keys "factory" and "manifest". Got: ${pluginPackage}`
        );
      }

      plugins.push({
        uri: uris[plugin],
        plugin: pluginPackage,
      });
    }

    if (config) {
      return new Web3ApiClient({
        ...config,
        plugins: [...plugins, ...(config.plugins ? config.plugins : [])],
      });
    } else {
      return new Web3ApiClient({ plugins });
    }
  }
);

File: packages/py/client/src/create_web3_api_client.py
from __future__ import annotations
import inspect

from .web3_api_client import Web3ApiClient, Web3ApiClientConfig
from .plugin_configs import PluginConfigs, modules, uris

async def create_web3_api_client(
    plugin_configs: PluginConfigs,
    config: Web3ApiClientConfig = None
) -> Web3ApiClient:
    plugins = []
    plugin_conf_dict = plugin_configs.__dict__ if plugin_configs else {}
    for plugin in plugin_conf_dict:
        if not modules.get(plugin, None):
            raise ValueError(f"Requested plugin \"{plugin}\" is not a supported createWeb3ApiClient plugin.")

        try:
            plugin_module = __import__(modules[plugin])
        except Exception as e:
            raise ValueError(f'Failed to import plugin module. Please install the package "{modules[plugin]}".\n Error: {str(e)}')

        plugin_factory = plugin_module["plugin"]

        if not plugin_factory:
            raise ValueError(f'Plugin module "{modules[plugin]}" is missing the "plugin: PluginFactory" export.')

        if not callable(plugin_factory):
            raise ValueError(f'The "plugin: PluginFactory" export must be a function. Found in module "{modules[plugin]}".')

        plugin_package = plugin_factory(plugin_configs)

        if (
            not plugin_package or
            not inspect.isclass(plugin_package) or
            not plugin_package.factory or
            not plugin_package.manifest
        ):
            raise ValueError(f'Plugin package is malformed. Expected object with keys "factory" and "manifest". Got: {plugin_package}')

        plugins.append({
            "uri": uris[plugin],
            "plugin": plugin_package
        })

    if config:
        return Web3ApiClient(
            config = Web3ApiClientConfig(
                redirects=config.redirects,
                plugins=plugins + config.plugins if config.plugins else [],
                interfaces=config.interfaces,
                envs=config.envs,
                uri_resolvers=config.uri_resolvers
            )
        )
    else:
        return Web3ApiClient(
            config = Web3ApiClientConfig(
                plugins=plugins
            )
        )
